### PR TITLE
RobotConstants and WheelConstants Dependency Injected into Firmware and Software

### LINKED ISSUE #2066: Create a `RobotConstants` struct to Initialize `FirmwareRobot`s
### Description of the task
Robot dependent constants like max_allowable_linear_acceleration, max_allowable_linear_speed, max_allowable_angular_acceleration, and max_allowable_angular_speed, JERK_LIMIT, WHEEL_MOTOR_PHASE_RESISTANCE, etc. are robot version dependent, so we need to be able to set these independently for each robot. This struct should be passed in to the constructor of FirmwareRobot and the values should replace uses of the constants file

### Acceptance criteria

- [ ] Create a `RobotConstants` struct
- [ ] initialize `FirmwareRobot` with this struct
- [ ] remove the associated constants from the constants file

### PR DESCRIPTION
**Pull request by jonathanlew**
### Description
* created RobotConstants struct with creator method
* created WheelConstants struct with creator method
* removed constants from constants.h and dependency injected the structs into software and firmware
* added RobotConstants to the Robot class

### Testing Done
added more unit tests on sharedphysics conversion function

### Resolved Issues
resolves #2066

### Length Justification
:(

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/firmware/shared/physics_linear_algebra.h
Comment (akhilveeraghanta) [inline src/firmware/shared/physics_linear_algebra.h:9]: looks like this file only has matrix operations, probably can just toss it in `matrix.h`?

so it becomes `shared_matrix_multiply()` and `shared_matrix_transpose()`?

File: src/shared/2015_robot_constants.c
Comment (akhilveeraghanta) [inline src/shared/2015_robot_constants.c]: any reason why these are defines and not just initialized directly in the struct?

File: src/software/ai/hl/stp/tactic/chip/chip_fsm_test.cpp
Comment (akhilveeraghanta) [inline src/software/ai/hl/stp/tactic/chip/chip_fsm_test.cpp]: curious why this chagned?

File: src/firmware/shared/physics.h
Comment (akhilveeraghanta) [inline src/firmware/shared/physics.h:55]: glad Jon was poked!

File: src/software/test_util/equal_within_tolerance.h
Comment (akhilveeraghanta) [inline src/software/test_util/equal_within_tolerance.h]: :joy: this was really funny you got an audible laugh out of me, can probably keep one or the other

**Review by akhilveeraghanta — COMMENTED** [review]
very impressive! thank you for doing this

just a couple comments from me

File: src/shared/2015_wheel_constants.h
Comment (kelvinkoon) [inline src/shared/2015_wheel_constants.h]: Is there a difference when we refer to something as "Legacy" and "2015"?

**Review by kelvinkoon — CHANGES_REQUESTED** [review]
Not much to add beyond Akhil's comments. 🙏  your work on this PR.

File: src/software/ai/hl/stp/tactic/chip/chip_fsm_test.cpp
Comment (jonathanlew) [reply src/software/ai/hl/stp/tactic/chip/chip_fsm_test.cpp]: I reduced the size of the getbehindball triangle, since it wasn't kicking properly. This was due to rotation being slightly slower in simulation because the radius is slightly larger than before (0.085->0.09, i.e. use max radius instead of an out of date measure value)

File: src/shared/2015_wheel_constants.h
Comment (jonathanlew) [reply src/shared/2015_wheel_constants.h]: legacy and 2015 is the same for now (2015 is more specific so I think it's preferred)

File: src/software/test_util/equal_within_tolerance.h
Comment (jonathanlew) [reply src/software/test_util/equal_within_tolerance.h]: ah I meant to update this

File: src/software/ai/hl/stp/tactic/kick/kick_fsm_test.cpp
Comment (MathewMacDougall) [inline src/software/ai/hl/stp/tactic/kick/kick_fsm_test.cpp]: same comment as Akhil, why did this change?

File: src/firmware/shared/physics_linear_algebra.h
Comment (jonathanlew) [reply src/firmware/shared/physics_linear_algebra.h:9]: porting those functions to use the `Matrix_t` struct is out of scope

**Review by MathewMacDougall — COMMENTED** [review]
Nice. This ended up really clean, especially in the new firmware :+1: (yay better firmware architecture). I think putting the `RobotConstants` in the Robot class was a great idea and made the injection in `ai` code much cleaner (as opposed to passing each struct into everything).

I did notice you split the radius back out from the constants, which I have no real strong opinion on since it's very unlikely we'd ever have a robot not at the max radius, and now it should be much easier to make a proper `RobotConstant` in the future if needed.

Really just one nit the same as Akhil, but overall I have nothing to add.

File: src/firmware/shared/physics_linear_algebra.h
Comment (jonathanlew) [reply src/firmware/shared/physics_linear_algebra.h:9]: https://github.com/UBC-Thunderbots/Software/issues/2111

**Comment by jonathanlew** [comment]
> Nice. This ended up really clean, especially in the new firmware +1 (yay better firmware architecture). I think putting the `RobotConstants` in the Robot class was a great idea and made the injection in `ai` code much cleaner (as opposed to passing each struct into everything).
> 
> I did notice you split the radius back out from the constants, which I have no real strong opinion on since it's very unlikely we'd ever have a robot not at the max radius, and now it should be much easier to make a proper `RobotConstant` in the future if needed.
> 
> Really just one nit the same as Akhil, but overall I have nothing to add.

Seemed pretty intimidating to depedency inject robot radius everywhere so I didn't bother, also for the reason above that robot radius isn't likely to differ much from the max

File: src/software/ai/hl/stp/tactic/kick/kick_fsm_test.cpp
Comment (jonathanlew) [reply src/software/ai/hl/stp/tactic/kick/kick_fsm_test.cpp]: copied from above: I reduced the size of the getbehindball triangle, since it wasn't kicking properly. This was due to rotation being slightly slower in simulation because the radius is slightly larger than before (0.085->0.09, i.e. use max radius instead of an out of date measure value)

File: src/firmware/boards/legacy_robot_stm32f4/io/wheels.c
Comment (garethellis0) [inline src/firmware/boards/legacy_robot_stm32f4/io/wheels.c]: :confused: - why is this too much effort?

File: src/firmware/boards/legacy_robot_stm32f4/main.c
Comment (garethellis0) [inline src/firmware/boards/legacy_robot_stm32f4/main.c]: nit; duplicate include

File: src/firmware/shared/physics.h
Comment (garethellis0) [inline src/firmware/shared/physics.h]: This comment doesn't make a ton of sense as a description for this function. Seems like it was just copy pasted?

I'd suggest something like:
```
Convert a set of given wheel speeds into linear x,y, and angular rotation speed
```

File: src/firmware/shared/physics.h
Comment (garethellis0) [inline src/firmware/shared/physics.h:54]: We should add units for both params.

File: src/firmware/shared/physics.h
Comment (garethellis0) [inline src/firmware/shared/physics.h]: It's important to indicate how this is measured. Is it independent of wheel size?

File: src/firmware/shared/physics.h
Comment (garethellis0) [inline src/firmware/shared/physics.h]: Units? what's the orientation for positive/negative forces?

File: src/firmware/shared/physics.h
Comment (garethellis0) [inline src/firmware/shared/physics.h:66]: We should add units here.

File: src/firmware/shared/physics.h
Comment (garethellis0) [inline src/firmware/shared/physics.h]: nit; `angle` -> `absolute` angle for both `_deg` params

File: src/firmware/shared/physics.h
Comment (garethellis0) [inline src/firmware/shared/physics.h]:
``` suggestion
* Minimum difference between two angles
```
(units are specified on the `@param`)

File: src/firmware/shared/physics.h
Comment (garethellis0) [reply src/firmware/shared/physics.h]: nit; missing `[in]` for both `_deg` params

File: src/firmware/shared/physics.h
Comment (garethellis0) [inline src/firmware/shared/physics.h:80]: nit; missing `[in]` for these params. Same goes for a lot of functions in this file.

File: src/firmware/shared/physics_linear_algebra.h
Comment (garethellis0) [reply src/firmware/shared/physics_linear_algebra.h:9]: Probably at least worth filing a ticket. I've definitely noticed that there are lots of things (like `dot2D`) that should be moved to the math library.

File: src/firmware/shared/physics_test.cpp
Comment (garethellis0) [inline src/firmware/shared/physics_test.cpp:108]: love the round-trip testing, great idea.

File: src/shared/2015_wheel_constants.c
Comment (garethellis0) [inline src/shared/2015_wheel_constants.c]: nit; this could be `static const`

File: src/shared/2015_robot_constants.c
Comment (garethellis0) [inline src/shared/2015_robot_constants.c:22]: nit; this could be `static const`

File: src/shared/2015_wheel_constants.h
Comment (garethellis0) [reply src/shared/2015_wheel_constants.h]: Yeah, I coined the "legacy" term, but 2015 is better.

File: src/shared/constants.h
Comment (garethellis0) [inline src/shared/constants.h]: just a thought, we could remove `_IN_` from this variable name. Out-of-scope, your call

File: src/shared/robot_constants.h
Comment (garethellis0) [inline src/shared/robot_constants.h]: we should add units to this variable name

File: src/shared/robot_constants.h
Comment (garethellis0) [inline src/shared/robot_constants.h]: we should add units to this variable name

Comment (garethellis0) [reply src/shared/robot_constants.h]: same comment goes for the other members of this struct that are missing units.

File: src/shared/test_util/test_util.cpp
Comment (garethellis0) [inline src/shared/test_util/test_util.cpp:18]: nit; for a few functions we've added in this PR, `create` is a bit redundant, this could probably just be `mockRobotConstants` for example.

File: src/software/ai/hl/stp/tactic/chip/simulated_chip_tactic_test.cpp
Comment (garethellis0) [inline src/software/ai/hl/stp/tactic/chip/simulated_chip_tactic_test.cpp:70]: `ROBOT_MAX_RADIUS_METERS` isn't injected here? not a criticism, just trying to understand why not

Comment (garethellis0) [reply src/software/ai/hl/stp/tactic/chip/simulated_chip_tactic_test.cpp:70]: this happens in a few other places too

File: src/software/simulation/physics/physics_world_test.cpp
Comment (garethellis0) [inline src/software/simulation/physics/physics_world_test.cpp]: nit; double include (I'd do a pass over the whole PR, there are few other places we do this).

File: src/software/simulation/physics/physics_world_test.cpp
Comment (garethellis0) [inline src/software/simulation/physics/physics_world_test.cpp:24]: just thinking, is the idea that eventually we'll be able to remove the direct dependency of most things on `2015_robot_constants` and can instead just depend on some centralized interface for getting them?

File: src/software/simulation/physics_simulator_ball.h
Comment (garethellis0) [inline src/software/simulation/physics_simulator_ball.h]: We should add a `@tparam` to the jdoc explaining what `T` is.

File: src/software/simulation/standalone_simulator.h
Comment (garethellis0) [inline src/software/simulation/standalone_simulator.h]: What is `this robot`? This is a bit confusing since `this` usually refers to the instance this member function is a called on (so in this case, an instance of `StandaloneSimulator`, which is very much not a robot).

**Review by garethellis0 — COMMENTED** [review]
Seems pretty sane to me overall, nice to see this finally starting to be dependency injected :)

File: src/software/ai/hl/stp/tactic/chip/simulated_chip_tactic_test.cpp
Comment (jonathanlew) [reply src/software/ai/hl/stp/tactic/chip/simulated_chip_tactic_test.cpp:70]: it doesn't seem worth it since it's hard to do especially with obstacle generation

File: src/software/simulation/physics/physics_world_test.cpp
Comment (jonathanlew) [reply src/software/simulation/physics/physics_world_test.cpp]: robot_constants vs wheel_constants

File: src/firmware/boards/legacy_robot_stm32f4/main.c
Comment (jonathanlew) [reply src/firmware/boards/legacy_robot_stm32f4/main.c]: robot_constants vs wheel_constants

File: src/firmware/boards/legacy_robot_stm32f4/io/wheels.c
Comment (jonathanlew) [reply src/firmware/boards/legacy_robot_stm32f4/io/wheels.c]: actually it's not that bad...

File: src/firmware/shared/physics.h
Comment (jonathanlew) [reply src/firmware/shared/physics.h]: fixed

File: src/firmware/shared/physics.h
Comment (jonathanlew) [reply src/firmware/shared/physics.h]: fixed first nit, second nit isn't part of code conventions since they are primitive types

File: src/firmware/shared/physics.h
Comment (jonathanlew) [reply src/firmware/shared/physics.h]: fixed

File: src/shared/2015_wheel_constants.c
Comment (jonathanlew) [reply src/shared/2015_wheel_constants.c]: added

File: src/firmware/boards/legacy_robot_stm32f4/main.c
Comment (jonathanlew) [reply src/firmware/boards/legacy_robot_stm32f4/main.c]: I just went ahead and merged the two files

File: src/firmware/shared/physics.h
Comment (jonathanlew) [reply src/firmware/shared/physics.h:54]: it's unitless, added a comment

File: src/firmware/shared/physics.h
Comment (jonathanlew) [reply src/firmware/shared/physics.h]: added comment

File: src/firmware/shared/physics.h
Comment (jonathanlew) [reply src/firmware/shared/physics.h]: added

File: src/shared/constants.h
Comment (jonathanlew) [reply src/shared/constants.h]: this constant isn't even used so removed

File: src/shared/robot_constants.h
Comment (jonathanlew) [reply src/shared/robot_constants.h]: added

File: src/shared/robot_constants.h
Comment (jonathanlew) [reply src/shared/robot_constants.h]: added

File: src/software/simulation/physics_simulator_ball.h
Comment (jonathanlew) [reply src/software/simulation/physics_simulator_ball.h]: added

File: src/software/simulation/standalone_simulator.h
Comment (jonathanlew) [reply src/software/simulation/standalone_simulator.h]: fixed

File: src/shared/2015_robot_constants.c
Comment (jonathanlew) [reply src/shared/2015_robot_constants.c]: fixed

File: src/firmware/shared/physics.h
Comment (jonathanlew) [reply src/firmware/shared/physics.h:66]: added

File: src/shared/2015_robot_constants.c
Comment (jonathanlew) [reply src/shared/2015_robot_constants.c:22]: added

File: src/software/ai/hl/stp/tactic/chip/simulated_chip_tactic_test.cpp
Comment (garethellis0) [reply src/software/ai/hl/stp/tactic/chip/simulated_chip_tactic_test.cpp:70]: Interesting, what's the gain we get from dependency injecting in other places in that case? Does a similar argument not apply here? 

(this is definitely not blocking)

Comment (jonathanlew) [reply src/software/ai/hl/stp/tactic/chip/simulated_chip_tactic_test.cpp:70]: dependency injecting into sw is good for trajectory planning in sw + potential hybrid fleet, and these values seem like they are likely to change from one robot revision to the next.

File: src/shared/test_util/test_util.cpp
Comment (jonathanlew) [reply src/shared/test_util/test_util.cpp:18]: create avoids problems with starting function names with numbers... is that a good reason?

File: src/software/simulation/physics/physics_world_test.cpp
Comment (jonathanlew) [reply src/software/simulation/physics/physics_world_test.cpp:24]: no we want to keep direct dependency so we can change simulation vs software vs firmware independently

File: src/shared/2015_wheel_constants.h
Comment (kelvinkoon) [reply src/shared/2015_wheel_constants.h]: Does `shared_physics_legacySpeed4ToSpeed3` refer to the same thing? It feels a bit weird since everywhere else except there is referred to as 2015.

**Review by akhilveeraghanta — DISMISSED** [review]
lgtm :100:

File: src/software/ai/hl/stp/tactic/chip/simulated_chip_tactic_test.cpp
Comment (garethellis0) [reply src/software/ai/hl/stp/tactic/chip/simulated_chip_tactic_test.cpp:70]: Ah, so we're just using the static constants for sim for now?

Comment (jonathanlew) [reply src/software/ai/hl/stp/tactic/chip/simulated_chip_tactic_test.cpp:70]: what was the reason for having different constants?

Comment (MathewMacDougall) [reply src/software/ai/hl/stp/tactic/chip/simulated_chip_tactic_test.cpp:70]: For sim? I might be missing context here but DI is useful in sim mostly for testing. For most physics / sim implementation code we don't want tests to fail if we tweak values to make them more realistic for gameplay.

**Comment by akhilveeraghanta** [comment]
lgtm pending merge conflict fix

**Review by akhilveeraghanta — DISMISSED** [review]

**Review by akhilveeraghanta — DISMISSED** [review]

**Review by kelvinkoon — APPROVED** [review]
LGTM pending CI!

**Review by akhilveeraghanta — APPROVED** [review]

**Pull request merged**
